server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedDumbBlobStore.java

```diff
@@ -0,0 +1,138 @@
+
+        return Mono.just(bucketName)
+            .filter(defaultBucket::equals)
+            .flatMap(bucker ->
+                Mono.from(cache.read(blobId))
+                    .<InputStream>flatMap(bytes -> Mono.fromCallable(() -> new ByteArrayInputStream(bytes)))
+                    .switchIfEmpty(Mono.fromCallable(() -> backend.read(bucketName, blobId)))
+            )
+            .switchIfEmpty(Mono.fromCallable(() -> backend.read(bucketName, blobId)))
+            .blockOptional()
+            .orElseThrow(() -> new ObjectNotFoundException(String.format("Could not retrieve blob metadata for %s", blobId)));
+    }
+
+    @Override
+    public Publisher<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
+        return Mono.from(cache.read(blobId))
+            .switchIfEmpty(Mono.from(backend.readBytes(bucketName, blobId)));
+    }
+
+    @Override
+    public Publisher<Void> save(BucketName bucketName, BlobId blobId, byte[] data) {
+        return Mono.from(backend.save(bucketName, blobId, data))
+            .then(Mono.just(bucketName)
+                .filter(bucket -> bucket.equals(defaultBucket) && data.length <= sizeThresholdInBytes)
+                .flatMap(ignored -> Mono.from(cache.cache(blobId, data))));
+    }
+
+    @Override
+    public Publisher<Void> save(BucketName bucketName, BlobId blobId, InputStream inputStream) {
+        return Mono.from(backend.save(bucketName, blobId, inputStream))
+            .then(Mono.just(inputStream)
+                .flatMap(stream -> handleCache(bucketName, blobId, stream)));
+    }
+
+    @Override
+    public Publisher<Void> save(BucketName bucketName, BlobId blobId, ByteSource content) {
+        return Mono.from(backend.save(bucketName, blobId, content))
+            .then(Mono.fromCallable(content::openBufferedStream)
+                .flatMap(stream -> handleCache(bucketName, blobId, stream)));
+    }
+
+    @Override
+    public Publisher<Void> delete(BucketName bucketName, BlobId blobId) {
+        return Mono.from(backend.delete(bucketName, blobId))
+            .then(Mono.from(cache.remove(blobId)));
+    }
+
+    @Override
+    public Publisher<Void> deleteBucket(BucketName bucketName) {
+        return Mono.from(backend.deleteBucket(bucketName))
+            .then(Mono.just(bucketName)
+                .filter(defaultBucket::equals)
+                .map(ignored ->
+                    bucketDAO.listAll()
+                        .filter(bucketNameBlobIdPair -> bucketNameBlobIdPair.getKey().equals(bucketName))
+                        .map(Pair::getValue)
+                        .flatMap(cache::remove)
+                ).then());
+    }
+
+    private Mono<Void> handleCache(BucketName bucketName, BlobId blobId, InputStream inputStream) {
+        return Mono.fromCallable(() -> {
+            byte[] bytes = new byte[sizeThresholdInBytes + 1];
+            PushbackInputStream pushbackInputStream = new PushbackInputStream(inputStream);
```

Comment:
it's too late, you can't push back data of the underlying inputstream, you have to wrap it here https://github.com/linagora/james-project/pull/3295/files/518c4e3745e67ea798ddb89d0a89bb78c4fb6531..64b4672abd46f0d1d7a5a770bce5d2246966da2b#diff-e1a0f91a59f3179a1c767bd27ea3953aR95